### ISSUE TITLE
TextField with input formatter does not handle backspace correctly

### ISSUE THREAD

**Issue by ppiatkowski** [flutter/flutter#68892]
## Steps to Reproduce

1. Run `flutter create textfield`.
2. In main.dart change build method of _MyHomePageState to look like below:
```
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Container(
        margin: EdgeInsets.all(15),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: <Widget>[
            Text(
              'TextField below uses LengthLimitingTextInputFormatter(5) input formatter',
              style: TextStyle(fontSize: 12),
            ),
            Text(
              '\nTry to enter 8 characters and then press backspace. You have to tap backspace'
              ' 4 times to delete the 5th character. '
              'Presumably first 3 taps delete the invisible characters over the limit',
              style: TextStyle(fontSize: 12),
            ),
            TextField(
              inputFormatters: <TextInputFormatter>[
                LengthLimitingTextInputFormatter(5)
              ],
            ),
            SizedBox(height: 25),
            Text(
              'TextField below uses FilteringTextInputFormatter.digitsOnly input formatter',
              style: TextStyle(fontSize: 12),
            ),
            Text(
              '\nTry to enter "123asd" and then press backspace. You have to tap backspace'
              ' 4 times to delete "3"'
              'Presumably first 3 taps delete "asd" characaters that were blocked by the input formatter',
              style: TextStyle(fontSize: 12),
            ),
            TextField(
              inputFormatters: <TextInputFormatter>[
                FilteringTextInputFormatter.digitsOnly
              ],
            ),
          ],
        ),
      ),
    );
  }
```

3. In the first Text Field enter any 8 characters. LengthLimitingTextInputFormatter will allow you to enter only 5 characters but keep adding more than 5. Then tap backspace. 
**Expected results:** 5th character gets deleted. 
**Actual results:** Nothing happens until you hit backspace 4th time. Presumably first 3 taps 'delete' 3 invisible characters blocked by the input formatter.

4. In the second Text Field enter '123asd'. FilteringTextInputFormatter.digitsOnly will allow you to enter only digits but keep tap 'a', 's', 'd' or any other non-digit characters. Then tap backspace. 
**Expected results:** '3' is deleted.
**Actual results:** Nothing happens until you hit backspace 4th time. Presumably first 3 taps 'delete' 3 invisible characters blocked by the input formatter.

```
✗ flutter analyze
Analyzing textfield...
No issues found! (ran in 2.9s)
```

```
➜  ✗ flutter doctor -v
[✓] Flutter (Channel stable, 1.22.2, on Mac OS X 10.15.7 19H2, locale en-PL)
    • Flutter version 1.22.2 at /Users/pawel/bin/flutter
    • Framework revision 84f3d28555 (8 days ago), 2020-10-15 16:26:19 -0700
    • Engine revision b8752bbfff
    • Dart version 2.10.2

[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
    • Android SDK at /Users/pawel/Library/Android/sdk
    • Platform android-29, build-tools 29.0.3
    • ANDROID_HOME = /Users/pawel/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.5)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.5, Build version 11E608c
    • CocoaPods version 1.9.3

[!] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.50.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.15.1

[✓] Connected device (2 available)
    • Android SDK built for x86 64 (mobile) • emulator-5554                        • android-x64 • Android 10 (API 29) (emulator)
    • iPhone SE (2nd generation) (mobile)   • B53EBEBC-3FE5-4781-A1E4-ED5CD3F4851E • ios         •
      com.apple.CoreSimulator.SimRuntime.iOS-13-5 (simulator)

! Doctor found issues in 1 category
```

</details>

**Comment by chinmoy12c** [comment]
I am not able to reproduce this bug. It works fine on my device.

<details>

<summary>flutter doctor -v</summary>

```

[✓] Flutter (Channel master, 1.24.0-2.0.pre.13, on Linux, locale en_US.utf8)
    • Flutter version 1.24.0-2.0.pre.13 at /home/Chinmoy/MasterFiles/openSource/flutter
    • Framework revision 2321b72dfb (8 days ago), 2020-10-15 09:20:52 -0700
    • Engine revision b22809b084
    • Dart version 2.11.0 (build 2.11.0-222.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.1)
    • Android SDK at /home/Chinmoy/Android/Sdk
    • Platform android-30, build-tools 30.0.1
    • Java binary at: /opt/android-studio/jre/bin/java
    • Java version openjdk version "1.8.0_242-release"
    • All Android licenses accepted.

[✓] Android Studio (version 3.4)
    • Android Studio at /home/Chinmoy/Downloads/android-studio
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version openjdk version "1.8.0_152-release"

[✓] Android Studio (version 4.0)
    • Android Studio at /opt/android-studio
    • Flutter plugin version 49.0.2
    • Dart plugin version 193.7547
    • Java version openjdk version "1.8.0_242-release"

[✓] VS Code (version 1.48.0)
    • VS Code at /usr/share/code
    • Flutter extension version 3.14.1

[✓] Connected device (1 available)
    • SM M215F (mobile) • RZ8N501AW9E • android-arm64 • Android 10 (API 29)

• No issues found!
```
![ezgif com-video-to-gif](https://user-images.githubusercontent.com/57670338/97049198-e8cea080-1598-11eb-8837-32b6c836e241.gif)
![ezgif com-video-to-gif](https://user-images.githubusercontent.com/57670338/97049223-f5eb8f80-1598-11eb-9973-2860575429c5.gif)
![ezgif com-video-to-gif](https://user-images.githubusercontent.com/57670338/97049324-2b907880-1599-11eb-8b15-7757f0c95205.gif)
``

</details>

<details>
<summary> main.dart file </summary>

```
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {

    Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Container(
        margin: EdgeInsets.all(15),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: <Widget>[
            Text(
              'TextField below uses LengthLimitingTextInputFormatter(5) input formatter',
              style: TextStyle(fontSize: 12),
            ),
            Text(
              '\nTry to enter 8 characters and then press backspace. You have to tap backspace'
              ' 4 times to delete the 5th character. '
              'Presumably first 3 taps delete the invisible characters over the limit',
              style: TextStyle(fontSize: 12),
            ),
            TextField(
              inputFormatters: <TextInputFormatter>[
                LengthLimitingTextInputFormatter(5)
              ],
            ),
            SizedBox(height: 25),
            Text(
              'TextField below uses FilteringTextInputFormatter.digitsOnly input formatter',
              style: TextStyle(fontSize: 12),
            ),
            Text(
              '\nTry to enter "123asd" and then press backspace. You have to tap backspace'
              ' 4 times to delete "3"'
              'Presumably first 3 taps delete "asd" characaters that were blocked by the input formatter',
              style: TextStyle(fontSize: 12),
            ),
            TextField(
              inputFormatters: <TextInputFormatter>[
                FilteringTextInputFormatter.digitsOnly
              ],
            ),
          ],
        ),
      ),
    );
  }
}

```

</details>

Although I'm facing a different bug in the ```LengthLimitingTextInputFormatter```. 

![ezgif com-video-to-gif](https://user-images.githubusercontent.com/57670338/97049414-4cf16480-1599-11eb-88be-dbb3be3430a0.gif)

**Comment by sidrao2006** [comment]
Hey @ppiatkowski ,
I found an issue addressing this problem - #67236. It was fixed in #67892. You could wait for it to be released to stable or upgrade to master.
Thanks

**Comment by shihaohong** [comment]
Since the issue has been resolved with #67892, I'll close this issue. If it turns out that this is a different issue, feel free to reopen.

**Comment by fbanyai** [comment]
Hi there,

I found a scenario where this bug is still seems to be happening...

When I have a field with `TextInputType.visiblePassword` and set my Capslock to ON the backspace doesn't work. If I set the Capslock to OFF then the backspace works as expected again.

**This bug was only when using Gboard as the keyboard. It didn't happen with Swype + Dragon.**

![20210429_090401](https://user-images.githubusercontent.com/2895008/116548904-3bcc3c00-a8cb-11eb-92f9-03fbd4159ff5.gif)

- Flutter (Channel beta, 2.0.2, on Microsoft Windows [Version 10.0.19041.928], locale en-US)
